src/number_theory/legendre_symbol/quadratic_char.lean

```diff
@@ -0,0 +1,331 @@
+  let sq : F → F := λ x, x^2,
+  have h : ¬ function.injective sq,
+  begin
+    simp only [function.injective, not_forall, exists_prop],
+    use [-1, 1],
+    split,
+    { simp only [sq, one_pow, neg_one_sq], },
+    { exact finite_field.neg_one_ne_one_of_char_ne_two hF, },
+  end,
+  have h₁ := mt (fintype.injective_iff_surjective.mpr) h, -- sq not surjective
+  push_neg at h₁,
+  cases h₁ with a h₁,
+  use a,
+  simp only [is_square, sq, not_exists, ne.def] at h₁ ⊢,
+  intros b hb,
+  rw ← pow_two at hb,
+  exact (h₁ b hb.symm),
+end
+
+end finite_field
+
+end general
+
+namespace char
+
+/-!
+### Definition of the quadratic character
+
+We define the quadratic character of a finite field `F` with values in ℤ.
+-/
+
+section define
+
+/-- Define the quadratic character with values in ℤ on a monoid with zero `α`.
+It takes the value zero at zero; for non-zero argument `a : α`, it is `1`
+if `a` is a square, otherwise it is `-1`.
+
+This only deserves the name "character" when it is multiplicative,
+e.g., when `α` is a finite field. See `quadratic_char_mul`.
+-/
+def quadratic_char (α : Type*) [monoid_with_zero α] [decidable_eq α]
+  [decidable_pred (is_square : α → Prop)] (a : α) : ℤ :=
+if a = 0 then 0 else if is_square a then 1 else -1
+
+end define
+
+/-!
+### Basic properties of the quadratic character
+
+We prove some properties of the quadratic character.
+We work with a finite field `F` here.
+The interesting case is when the characteristic of `F` is odd.
+-/
+
+section quadratic_char
+
+variables {F : Type*} [field F] [fintype F] [decidable_eq F]
+
+/-- Some basic API lemmas -/
+lemma quadratic_char_eq_zero_iff (a : F) : quadratic_char F a = 0 ↔ a = 0 :=
+begin
+  simp only [quadratic_char],
+  by_cases ha : a = 0,
+  { simp only [ha, eq_self_iff_true, if_true], },
+  { simp [ha],
+    split_ifs; simp only [neg_eq_zero, one_ne_zero, not_false_iff], },
+end
+
+@[simp]
+lemma quadratic_char_zero : quadratic_char F 0 = 0 :=
+by simp only [quadratic_char, eq_self_iff_true, if_true, id.def]
+
+@[simp]
+lemma quadratic_char_one : quadratic_char F 1 = 1 :=
+by simp only [quadratic_char, one_ne_zero, is_square_one, if_true, if_false, id.def]
+
+/-- For nonzero `a : F`, `quadratic_char F a = 1 ↔ is_square a`. -/
+lemma quadratic_char_one_iff_is_square {a : F} (ha : a ≠ 0) :
+  quadratic_char F a = 1 ↔ is_square a :=
+by { simp [quadratic_char, ha, (dec_trivial : (-1 : ℤ) ≠ 1)], tauto }
+
+/-- The quadratic character takes the value `1` on nonzero squares. -/
+lemma quadratic_char_sq_one' {a : F} (ha : a ≠ 0) : quadratic_char F (a ^ 2) = 1 :=
+by simp only [quadratic_char, ha, pow_eq_zero_iff, nat.succ_pos', is_square_sq, if_true, if_false]
+
+/-- If `ring_char F = 2`, then `quadratic_char F` takes the value `1` on nonzero elements. -/
+lemma quadratic_char_eq_one_of_char_two (hF : ring_char F = 2) {a : F} (ha : a ≠ 0) :
+  quadratic_char F a = 1 :=
+begin
+  simp [quadratic_char, ha],
```

Comment:
This is a non-terminal simp (see https://leanprover-community.github.io/glossary.html#non-terminal-simp). Please squeeze it.

Reply:
OK.